src/main/java/com/fasterxml/jackson/databind/deser/BeanDeserializerBase.java

```diff
@@ -1902,12 +1907,17 @@ protected JsonDeserializer<Object> _findSubclassDeserializer(DeserializationCont
         subDeser = ctxt.findRootValueDeserializer(type);
         // Also, need to cache it
         if (subDeser != null) {
-            synchronized (this) {
-                if (_subDeserializers == null) {
-                    _subDeserializers = new HashMap<ClassKey,JsonDeserializer<Object>>();;
+            if (_subDeserializers == null) {
```

Comment:
Check for `null` is not guarded so I am not sure this is actually safe?

Reply:
_subDeserializers == null is checked again after the lock is made.

This is typical of optimistic state management.
* check state
* if state is not what you want, acquire lock - check state again and then update the state if needed
* if state is already what you want, then just proceed without needing to lock

Reply:
Ok. So even tho without sync memory view wrt null-to-non-null transition might not be visible, locking is a memory barrier at which view is updated so it will all work because it is one-way transition. Makes sense.